---
title: "HW3"
author: "Kelsey Grinde"
date: "3/2/2015"
output: html_document
---

# Instructions:
The goal of this homework is to reproduce the results in Figure 2 in this paper: Qian, F., Bolen, C. R., Jing, C., Wang, X., Zheng, W., Zhao, H., et al. (2013). Impaired toll-like receptor 3-mediated immune responses from macrophages of patients chronically infected with hepatitis C virus. Clinical and Vaccine Immunology : CVI, 20(2), 146â€“155. doi:10.1128/CVI.00530-12


First, a little setup to get all the packages we will need.
```{r set-up}
# setup for use of Bioconductor
suppressMessages(source('http://bioconductor.org/biocLite.R'))

# download packages
suppressMessages(library(GEOmetadb))
suppressMessages(library(limma))
suppressMessages(library(Biobase))
suppressMessages(library(data.table))
```

# Getting the data
Now we're ready to get the data from GEO:

```{r get-data,cache=T}
# get the data, accession number GSE40812
# This accession number is stated in the paper, but we also could have used our results from HW2 to figure out which accession number to use
data <- getGEO("GSE40812", destdir = getwd())

# pull out the data
data <- data[[1]]
```


We are going to need to clean up the information about the samples a bit. There is some information here that we really don't need for this analysis, and other information that can be abbreviated to make things a little cleaner.
```{r get-samp-info}
sample_info <- pData(data) # data.frame with one row per sample (80)
```

```{r explore-pData, echo = F, eval =F}
## This code is used to look at the sample information and figure out how to clean it

### geo_accession is same as row names
### title may be helpful to keep
### for source_name_ch1 we have two options (keep but clean):
levels(sample_info$source_name_ch1)
### for characteristics_ch1 we have two options (keep but clean):
levels(sample_info$characteristics_ch1)
### for characteristics_ch1.1 we have two options (keep but clean):
levels(sample_info$characteristics_ch1.1)
### for characteristics_ch1.2 we have three options (keep but clean):
levels(sample_info$characteristics_ch1.2)
### for description we have 8 options (keep but clean):
levels(sample_info$description)
### don't need anything else (all the same)
```

```{r clean-sample-info}
# cleaning up the pData
keepCols <- c('source_name_ch1','characteristics_ch1','characteristics_ch1.1',
              'characteristics_ch1.2','description','title')
pD <- sample_info[,keepCols]
pD <- within(pD, {
  cell_type <- ifelse(source_name_ch1=='Monocyte-derived Macrophage','Macrophage','PBMC')
  infected <- ifelse(characteristics_ch1=='infection status: Pos',"pos","neg")
  treatment <- ifelse(characteristics_ch1.2=='treatment: Mock','Mock', 'Poly')
  subj_id <- substr(title,regexpr("_",title)+1,regexpr("_",title)+4)
  }
)
pD$characteristics_ch1.2 <- NULL
pD$source_name_ch1 <- NULL
pD$characteristics_ch1 <- NULL
pD$characteristics_ch1.1 <- NULL
pD$description <- NULL
pD$title <- NULL
```

Now, we notice that there are two different experiements here: one involving macrophages and one involving PBMCs. To reproduce Figure 2 we only need the macrophage data.
```{r get-only-macrophages}
# keep only the macrophage data
pD <- subset(pD, cell_type=="Macrophage")
pData(data) <- pD

# get the expression data
eSet <- exprs(data)
# only keep the macrophage data
eSet <- eSet[,rownames(pD)]

# feature data including gene name, chromosome and much more
feature_info <- fData(data) # data.frame with one row per feature
### pull out just ID, ILMN_Gene, Chromosome
features <- feature_info[,c("ID","ILMN_Gene","Chromosome")]
```

# Normalizing the data
The next step is to normalize the data (if necessary)

According to the paper, the data have already been normalized using quantile normalization via the `beadarray` package in bioconductor, so there is nothing more for us to do here. We can also check whether the data have been normalized by looking at the sample information we extracted using `pData()` (note that here it says the data were normalized using the `lumi` package in bioconductor).

```{r data-processing}
# print out data processing information
levels(sample_info$data_processing)
```

# Using LIMMA
Now we will use LIMMA to test for differential expression.

First the authors looked for genes that were differentially expressed by macrophages, looking at changes between Mock and Poly IC H for each subject, controlling FDR at 5% using the BH method and also restricting themselves to looking at genes with a 1.5-fold change in expression.

```{r de-tx, cache=T}
# look at differential expression between mock and poly, adjusting for infection
mm1 <- model.matrix(~treatment,pD)

# fit LIMMA model
fit1 <- lmFit(eSet,mm1)
ebay1 <- eBayes(fit1)

# get genes where treatment is signifcant and we see at least a 1.5-fold change
topTx1 <- topTable(ebay1,coef='treatmentPoly',adjust.method='BH',p.value = 0.05,lfc=log2(1.5),number=Inf)
sigGenes <- rownames(topTx1)
length(sigGenes) # 1146, matches the paper

##########################################
##########################################
### alternatively, we might want to include subject since we have paired samples
mm1_alt <- model.matrix(~subj_id+treatment,pD)

# fit LIMMA model
fit1_alt <- lmFit(eSet,mm1_alt)
ebay1_alt <- eBayes(fit1_alt)

# get genes where treatment is signifcant and we see at least a 1.5-fold change
topTx1_alt <- topTable(ebay1_alt,coef='treatmentPoly',adjust.method='BH',p.value = 0.05,lfc=log2(1.5),number=Inf)
sigGenes_alt <- rownames(topTx1_alt)
length(sigGenes_alt) # now we have 1153, a few more than the paper found
```

Further analyses focus only on these genes found above with significant differential expression between the two treatments. I'm going to use the 1146 genes found using the method described in the paper to see if I can reproduce their later results.

```{r restrict-to-deTx-genes}
# keep only significant genes and macrophage data
restrict_eSet <- exprs(data[sigGenes,rownames(pD)])
```

Next the authors calculated the fold changes between the Mock and Poly IC H samples for each subject.

```{r fold-change}
# get log fold change from mock to treatment for each subject for each gene
mm2 <- model.matrix(~0+subj_id*treatment,pD)

# clean up column names so we can make contrasts
colnames(mm2) <- substr(colnames(mm2),6,nchar(colnames(mm2)))
colnames(mm2)[21] <- 'id2014_Poly'
colnames(mm2)[22:40] <- gsub(':treatment','_',colnames(mm2)[22:40])

# get expression for each subject in mock and treatment
fit2 <- lmFit(restrict_eSet,mm2) # run just on restricted eSet

# make contrasts to get change for each subject from mock to treatment
cont2_mat <- makeContrasts(id2014_Poly-id2014,id2046_Poly-id2046,id2047_Poly-id2047,
                           id2048_Poly-id2048,id2049_Poly-id2049,id2050_Poly-id2050,
                           id2051_Poly-id2051,id2052_Poly-id2052,id2053_Poly-id2053,
                           id2054_Poly-id2054,id3049_Poly-id3049,id3050_Poly-id3050,
                           id3052_Poly-id3052,id3053_Poly-id3053,id3054_Poly-id3054,
                           id3055_Poly-id3055,id3056_Poly-id3056,id3057_Poly-id3057,
                           id3058_Poly-id3058,id3059_Poly-id3059,
                           levels=colnames(mm2))

cont_fit2 <- contrasts.fit(fit2,cont2_mat) # this should be estimated lfc for each subject between their mock and poly samples at each gene

# get new response for next analysis:
lfc_subj <- cont_fit2$coef
colnames(lfc_subj) <- c('id2014','id2046','id2047','id2048','id2049','id2050',
                        'id2051','id2052','id2053','id2054','id3049','id3050',
                        'id3052','id3053','id3054','id3055','id3056','id3057',
                        'id3058','id3059')

# figure out which subject is infected vs not
subj_info <- pD[,c('subj_id','infected')]
subj_info <- subj_info[order(subj_info $subj_id),]
df_infected <- subj_info[seq(1,nrow(subj_info),2),]
rownames(df_infected) <- paste('id',df_infected$subj_id,sep='')
df_infected$subj_id <- NULL

```

They compared these fold changes for HCV + and HCV - subjects to see if there were significantly different fold changes between HCV + and HCV - subjects.

```{r compare-lfc}
mm3 <- model.matrix(~infected,df_infected)

# fit LIMMA model
fit3 <- lmFit(lfc_subj,mm3)
ebay3 <- eBayes(fit3)

# look at differential lfc between HCV + vs HCV - subjects
topT3 <- topTable(ebay3,coef='infectedpos',adjust.method='BH',p.value=0.05,number=Inf) # nothing significant after BH adjustment

# try again with no BH adjustment
topT3_noBH <- topTable(ebay3, coef= 'infectedpos',adjust.method='none',p.value=0.1,lfc=log2(1.5), number=Inf)
```

No genes show significant differences beteween HCV + and HCV - subjects after adjusting p-values to control FDR at 5% via BH, but I find that 30 have an unadjusted p-value less than 0.1. The author found 43, and the differences could be due to the way I analyzed the data above.

# Making a heatmap
Display the results using a heatmap [Hint: Use the pheatmap package]

```{r}
library(pheatmap)
# get the probes from the last analysis
keep_probes <- rownames(topT3_noBH)
pheatmap_eSet <- restrict_eSet[rownames(restrict_eSet) %in% keep_probes,]

# get pD for heatmap
keep_pD = pD[,c("subj_id","treatment","infected")]

#Sort by mock vs poly then by HCV- vs HCV + and then by ID
keep_pD = keep_pD[order(keep_pD[,"treatment"],keep_pD[,"infected"],keep_pD[,'subj_id']),]

# resort as above
pheatmap_eSet <- pheatmap_eSet[,rownames(keep_pD)]

# change column names to subject IDs
colnames(pheatmap_eSet) <- keep_pD[,'subj_id']

# now display a heatmap of the expression values
pheatmap(mat=pheatmap_eSet,cluster_rows=FALSE,cluster_cols=FALSE)
```
